### PR TITLE
[FSDP2] Allocated reduce-scatter output in default stream

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132445

This PR allocates the reduce-scatter output in the default stream to reduce fragmentation. The main tricky part is that the reduce-scatter output experiences different code paths: e.g. it can be upcast to the original parameter dtype if needed, it may need to be all-reduced for HSDP, it may need to be used for sharded gradient accumulation after reduction, etc.

The simplest and still effective approach then is to simply hold onto the reduce-scatter output tensor until the end of backward where we have the default stream wait on the post-reduce event anyway.

---

**(Before)** Llama3-8B, 1D FSDP, 8 H100s, bf16/fp32 mixed precision, no AC, local batch size 1: (this is the "After" from https://github.com/pytorch/pytorch/pull/129721)
```
[rank0]:2024-06-27 16:41:12,106 - root - INFO - step:  1  loss: 12.2560  memory: 69.46GiB(73.08%)  wps: 1,158  mfu: 6.78%
[rank0]:2024-06-27 16:41:12,106 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:2024-06-27 16:41:13,502 - root - INFO - step:  2  loss: 12.0949  memory: 77.29GiB(81.32%)  wps: 5,870  mfu: 34.37%
[rank0]:2024-06-27 16:41:14,839 - root - INFO - step:  3  loss: 11.7770  memory: 77.29GiB(81.32%)  wps: 6,130  mfu: 35.90%
[rank0]:2024-06-27 16:41:16,154 - root - INFO - step:  4  loss: 11.3188  memory: 77.29GiB(81.32%)  wps: 6,230  mfu: 36.48%
[rank0]:2024-06-27 16:41:17,474 - root - INFO - step:  5  loss: 10.9443  memory: 77.29GiB(81.32%)  wps: 6,211  mfu: 36.37%
```

**(After)** Llama3-8B, 1D FSDP, 8 H100s, bf16/fp32 mixed precision, no AC, local batch size 1:
```
rank0]:2024-08-01 14:10:37,888 - root - INFO - step:  1  loss: 12.2337  memory: 70.67GiB(74.36%)  wps: 811  mfu: 4.75%
[rank0]:2024-08-01 14:10:37,891 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:2024-08-01 14:10:41,302 - root - INFO - step:  2  loss: 12.0674  memory: 74.58GiB(78.47%)  wps: 2,402  mfu: 14.07%
[rank0]:2024-08-01 14:10:42,619 - root - INFO - step:  3  loss: 11.7432  memory: 74.58GiB(78.47%)  wps: 6,224  mfu: 36.44%
[rank0]:2024-08-01 14:10:43,931 - root - INFO - step:  4  loss: 11.2658  memory: 74.58GiB(78.47%)  wps: 6,244  mfu: 36.57%
[rank0]:2024-08-01 14:10:45,243 - root - INFO - step:  5  loss: 10.9223  memory: 74.58GiB(78.47%)  wps: 6,246  mfu: 36.58%
```

Another 2.71 GiB reduction in reserved memory

---

This change looks like it may use more memory for `MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.bfloat16)` with fp32 sharded parameters. However, due to fragmentation effects, this is not the case for the particular Llama3-8B workload:

**(Before)** Llama3-8B, 1D FSDP, 8 H100s, bf16/fp32 mixed precision, no AC, local batch size 1, **bf16 reduce-scatter**:
```
[rank0]:2024-08-01 14:20:02,190 - root - INFO - step:  1  loss: 12.2899  memory: 70.85GiB(74.55%)  wps: 798  mfu: 4.67%
[rank0]:2024-08-01 14:20:02,190 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:2024-08-01 14:20:05,668 - root - INFO - step:  2  loss: 12.1181  memory: 78.68GiB(82.79%)  wps: 2,356  mfu: 13.79%
[rank0]:2024-08-01 14:20:09,238 - root - INFO - step:  3  loss: 11.7911  memory: 78.68GiB(82.79%)  wps: 2,295  mfu: 13.44%
[rank0]:2024-08-01 14:20:11,612 - root - INFO - step:  4  loss: 11.3347  memory: 78.68GiB(82.79%)  wps: 3,452  mfu: 20.21%
[rank0]:2024-08-01 14:20:12,897 - root - INFO - step:  5  loss: 10.9379  memory: 78.68GiB(82.79%)  wps: 6,376  mfu: 37.34%
``` 

**(After)** Llama3-8B, 1D FSDP, 8 H100s, bf16/fp32 mixed precision, no AC, local batch size 1, **bf16 reduce-scatter**:
```
[rank0]:2024-08-01 14:17:57,992 - root - INFO - step:  1  loss: 12.2500  memory: 70.55GiB(74.24%)  wps: 1,308  mfu: 7.66%
[rank0]:2024-08-01 14:17:57,992 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:2024-08-01 14:17:59,277 - root - INFO - step:  2  loss: 12.0963  memory: 78.38GiB(82.47%)  wps: 6,376  mfu: 37.34%
[rank0]:2024-08-01 14:18:00,567 - root - INFO - step:  3  loss: 11.7726  memory: 78.38GiB(82.47%)  wps: 6,357  mfu: 37.23%
[rank0]:2024-08-01 14:18:01,853 - root - INFO - step:  4  loss: 11.3399  memory: 78.38GiB(82.47%)  wps: 6,370  mfu: 37.30%
[rank0]:2024-08-01 14:18:03,137 - root - INFO - step:  5  loss: 10.9152  memory: 78.38GiB(82.47%)  wps: 6,386  mfu: 37.39%
```

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 1]
```diff
@@ -374,6 +378,7 @@ def post_backward(self, *unused: Any):
                 self.comm_ctx.reduce_scatter_state = None
             (
                 reduce_scatter_input,
+                self._reduce_output,
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:381]: this breaks optimizer in backward :( ugh, big pain

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:381]: and CPU offloading :(

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:381]: The issue is that there is not a good/simple point at which to have the default stream wait for the `post_reduce_event` in the middle of backward :/
